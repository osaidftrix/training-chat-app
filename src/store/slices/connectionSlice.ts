import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from '../../store';
import { HubConnection } from '@microsoft/signalr';

export interface IUser {
  username: string;
}
export interface IMessage extends IUser {
  message: string;
}
export interface IMessagePayload {
  target: string;
  message: IMessage;
}
export interface Message {
  [key: string]: IMessage[];
}

export interface ConnectionState {
  connectionInstance: HubConnection | null;
  isAuthenticated: boolean;
  userName: string | null;
  connectionId: string | null;
  messages: Message;
  loggedInUsers: IUser[];
}

export const initialState: ConnectionState = {
  connectionInstance: null,
  isAuthenticated: false,
  userName: null,
  connectionId: null,
  messages: {
    public: [],
  },
  loggedInUsers: [],
};

export const counterSlice = createSlice({
  name: 'connection',
  initialState,
  reducers: {
    setLoggedInUsers: (state, action: PayloadAction<IUser[]>) => {
      state.loggedInUsers = action.payload;
    },
    setMessages: (state, action: PayloadAction<IMessagePayload>) => {
      if (!state.messages[action.payload.target]) {
        state.messages[action.payload.target] = [action.payload.message];
      } else {
        state.messages[action.payload.target] = [...state.messages[action.payload.target], action.payload.message];
      }
    },
    setConnection: (state, action: PayloadAction<ConnectionState>) => {
      const { connectionInstance, isAuthenticated, userName } = action.payload;
      state.connectionInstance = connectionInstance;
      state.isAuthenticated = isAuthenticated;
      state.userName = userName;
      state.messages = action.payload.messages;
      state.connectionId = action.payload.connectionId;
    },
    setConnectionInstance: (state, action: PayloadAction<{ connectionInstance: HubConnection }>) => {
      state.connectionInstance = action.payload.connectionInstance;
      state.isAuthenticated = false;
    },
    removeConnection: (state) => {
      if (state.connectionInstance) {
        state.connectionInstance
          .stop()
          .then((result) => {
            console.log('connection disconnected: ', result);
          })
          .catch((err) => {
            console.log('Connection Removal Error: ', err);
          });
      }
      state.connectionInstance = null;
    },

    logOut: (state) => {
      state.connectionInstance?.invoke('Logout', state.userName);
      state.isAuthenticated = false;
      state.connectionId = null;
      state.userName = null;
      state.loggedInUsers = [];
      state.messages = { public: [] };
    },
  },
});

export const { setConnection, removeConnection, setLoggedInUsers, setMessages, logOut, setConnectionInstance } =
  counterSlice.actions;

export const getConnection = (state: RootState) => state.connection;

export default counterSlice.reducer;
