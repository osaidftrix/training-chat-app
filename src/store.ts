import { combineReducers, configureStore } from '@reduxjs/toolkit';
import connectionSlice from './store/slices/connectionSlice';
import storage from 'redux-persist/lib/storage';
//import storage from 'redux-persist-indexeddb-storage';
import { persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['connection'],
  // whitelist: ['connection']
};

const rootReducer = combineReducers({
  connection: connectionSlice,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: [thunk],
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
