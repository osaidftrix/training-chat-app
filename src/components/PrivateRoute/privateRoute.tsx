import { Navigate } from 'react-router-dom';
import { useAppSelector } from '../../store/hooks';
import { getConnection } from '../../store/slices/connectionSlice';
import Layout from '../Layout/layout';
import React, { ReactNode } from 'react';

interface IPrivateRoute {
    redirectPath: string;
    component: ReactNode; 
}
export const PrivateRoute: React.FC<IPrivateRoute> = ({ redirectPath, component }) => {
  const isAuthenticated = useAppSelector(getConnection).isAuthenticated;
  if (!isAuthenticated) return <Navigate to={redirectPath} />;
  return <Layout>{component}</Layout>;
};
