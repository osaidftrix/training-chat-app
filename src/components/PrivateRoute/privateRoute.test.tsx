import { describe, test } from 'vitest';
import { Provider } from 'react-redux';
import { persistor, store } from '../../store';
import { MemoryRouter, Routes, Route } from 'react-router';
import { PersistGate } from 'redux-persist/integration/react';
import { PrivateRoute } from './privateRoute';
import { render, screen } from '@testing-library/react';
import { initialState, setConnection } from '../../store/slices/connectionSlice';

describe('<PrivateRoute />', () => {
  test('PrivateRoute renders properly', () => {
    const { container } = render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/test']}>
            <Routes>
              <Route
                path="/test"
                element={
                  <PrivateRoute
                    redirectPath={'/'}
                    component={
                      <div>
                        <span>Authenticated</span>
                      </div>
                    }
                  />
                }
              ></Route>
              <Route
                path="/"
                element={
                  <div>
                    <span>Not Authenticated</span>
                  </div>
                }
              ></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    expect(container).toBeTruthy();
  });

  test('PrivateRoute redirects to default route if not authenticated', () => {
    render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/test']}>
            <Routes>
              <Route
                path="/test"
                element={
                  <PrivateRoute
                    redirectPath={'/'}
                    component={
                      <div>
                        <span>Authenticated</span>
                      </div>
                    }
                  />
                }
              ></Route>
              <Route
                path="/"
                element={
                  <div>
                    <span>Not Authenticated</span>
                  </div>
                }
              ></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    const element = screen.getByText('Not Authenticated');
    expect(element).toBeTruthy();
  });

  test('PrivateRoute does not redirect to default route if authenticated', () => {
    store.dispatch(
      setConnection({...initialState, isAuthenticated: true}),
    );
    render(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <MemoryRouter initialEntries={['/test']}>
            <Routes>
              <Route
                path="/test"
                element={
                  <PrivateRoute
                    redirectPath={'/'}
                    component={
                      <div>
                        <span>Authenticated</span>
                      </div>
                    }
                  />
                }
              ></Route>
              <Route
                path="/"
                element={
                  <div>
                    <span>Not Authenticated</span>
                  </div>
                }
              ></Route>
            </Routes>
          </MemoryRouter>
        </PersistGate>
      </Provider>,
    );
    const element = screen.getByText('Authenticated');
    expect(element).toBeTruthy();
  });
});
