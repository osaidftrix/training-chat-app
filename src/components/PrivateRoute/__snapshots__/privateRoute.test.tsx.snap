// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PrivateRoute /> > PrivateRoute renders properly 1`] = `null`;
